agent/qrexec-agent.c

```diff
@@ -120,6 +126,86 @@ static struct pam_conv conv = {
     NULL
 };
 #endif
+
+_Noreturn static void really_exec(const char *prog, const struct passwd *pw,
+                                  char **env, const char *cmd, pid_t agent_pid)
+{
+    /* 1 means init, which is completely wrong. */
+    if (agent_pid < 2) {
+        LOG(ERROR, "Bad agent_pid value (did parent process die?): %ju",
+            (uintmax_t)agent_pid);
+        _exit(QREXEC_EXIT_PROBLEM);
+    }
+
+    /* child */
+    setsid();
+
+    /* try to enter home dir, but don't abort if it fails */
+    int retval = chdir(pw->pw_dir);
+    if (retval == -1)
+        warn("chdir(%s)", pw->pw_dir);
+
+    /* call QUBESRPC if requested */
+    if (prog) {
+        /* Set up environment variables for a login shell. */
+        char *pid_str;
+        if (asprintf(&pid_str, "%ju", (uintmax_t)agent_pid) < 1) {
+            PERROR("asprintf");
+            _exit(QREXEC_EXIT_PROBLEM);
+        }
+        const char *buf[] = {"sh", "-lc", "exec \"$@\"", "sh", exec_program, pid_str, prog, cmd, service_path, NULL};
+        /* TODO: use the user's shell not /bin/sh */
```

Comment:
While it may sounds like a smart move, it's also problematic, as the script might need to be adjusted for such shell. AFAIR "user shell" doesn't even need to be POSIX compliant.